### ISSUE TITLE
There was an unknown error starting Kubernetes: FetchError: request to https://update.k3s.io/v1-release/channels failed, reason: certificate signature failure

### ISSUE THREAD

**Issue by drenze-athene** [rancher-sandbox/rancher-desktop#1261]
### Rancher Desktop Version

1.0.0-beta.1

### Rancher Desktop K8s Version

1.21.6

### What operating system are you using?

macOS

### Operating System / Build Version

macOS Big Sur 11.6.1

### What CPU architecture are you using?

x64

### Actual Behavior

When starting `v1.0.0-beta`, get the following error:

> There was an unknown error starting Kubernetes: FetchError: request to https://update.k3s.io/v1-release/channels failed, reason: certificate signature failure

This **only** appears to occur when connecting from inside the corporate network, which uses a self-signed cert for the internal proxy. When disconnected from the internal network, everything starts up as expected.

### Steps to Reproduce

1. Start `1.0.0-beta.1` for the first time after installation, while behind a firewall with a self-signed cert.
2. Receive the error shown above (and in the screenshot, below) prior to downloading Kubernetes.

### Result

![image](https://user-images.githubusercontent.com/81439124/149964611-e45641eb-5d6a-4e23-ae74-f8862a7cdcc3.png)

### Expected Behavior

Kubernetes should start.

**Comment by evertonlperes** [comment]
Hi @drenze-athene 
Thanks for reporting.

Could you also provide the logs? You can find the logs folder navigating to `Troubleshooting` tab and `Show logs`.

Thanks.

**Comment by drenze-athene** [comment]
> Hi @drenze-athene Thanks for reporting.
> 
> Could you also provide the logs? You can find the logs folder navigating to `Troubleshooting` tab and `Show logs`.
> 
> Thanks.

I'll do a clean install again and do so. Are there specific logs that you need, or all logs?

Additionally - once I do the initial install and download while connected to a network outside the corporate firewall, it will start up again just fine. I did a restart

**Comment by evertonlperes** [comment]
> > Hi @drenze-athene Thanks for reporting.
> > Could you also provide the logs? You can find the logs folder navigating to `Troubleshooting` tab and `Show logs`.
> > Thanks.
> 
> I'll do a clean install again and do so. Are there specific logs that you need, or all logs?
> 
> Additionally - once I do the initial install and download while connected to a network outside the corporate firewall, it will start up again just fine. I did a restart and it started without any issues.

Awesome.
Feel free to share all logs (some will be probably empty, you don't need to worry about those).

I see. That's make sense, once you have k3s binary into the cache folder, it won't required to access the internet to make it working, but if you need to upgrade/downgrade, there's a chance to face the same error (on corporate network).

Here's a couple more questions that will help us a lot :D

* Could you also inform which firewall/vpn tools are you using (any information about your setup will help a lot)?
* Could you share the logs after the RD installation, under the corporate network?

Thanks!

**Comment by drenze-athene** [comment]
> Awesome. Feel free to share all logs (some will be probably empty, you don't need to worry about those).
> 

The only log that has anythingin it is the `update.log`, which shows the following:

```
2022-01-18T15:41:12.495Z: Checking for update
2022-01-18T15:41:18.869Z: Error: FetchError: request to https://desktop.version.rancher.io/v1/checkupgrade failed, reason: certificate signature failure
    at ClientRequest.<anonymous> (/Applications/Rancher Desktop.app/Contents/Resources/app.asar/node_modules/node-fetch/lib/index.js:1483:11)
    at ClientRequest.emit (node:events:394:28)
    at TLSSocket.socketErrorListener (node:_http_client:447:9)
    at TLSSocket.emit (node:events:394:28)
    at emitErrorNT (node:internal/streams/destroy:157:8)
    at emitErrorCloseNT (node:internal/streams/destroy:122:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
```

> > Additionally - once I do the initial install and download while connected to a network outside the corporate firewall, it will start up again just fine. I did a restart and it started without any issues.
> 
> I see. That's make sense, once you have k3s binary into the cache folder, it won't required to access the internet to make it working, but if you need to upgrade/downgrade, there's a chance to face the same error (on corporate network).

Ya - I meant to add that this mirrors the experience I had when using Docker Desktop, prior to moving to Rancher Desktop. I would have this issue each time there was an upgrade, although with RD, I didn't have this issue prior to upgrading to `1.0.0-beta`.

> Here's a couple more questions that will help us a lot :D
> 
> * Could you also inform which firewall/vpn tools are you using (any information about your setup will help a lot)?
> * Could you share the logs after the RD installation, under the corporate network?
> 

Regarding your additional questions:

  * I'm on-site today, so no VPN.
  * Truly not certain what our firewall is.
  * MacOS installation is to unzip the file, then drag-and-drop into the application folder and start, not certain where to find any installation logs.

> Thanks!

YW.

**Comment by drenze-athene** [comment]
> > I see. That's make sense, once you have k3s binary into the cache folder, it won't required to access the internet to make it working, but if you need to upgrade/downgrade, there's a chance to face the same error (on corporate network).

So, I downgraded k3s from 1.23.x to 1.20.x to match what we're currently running in our cluster, and the downgrade did **not** require me to disconnect from the corporate network to do so. Obviously no word on an upgrade/downgrade for RD yet.

**Comment by evertonlperes** [comment]
Thanks for sharing @drenze-athene - it will help a lot.

Regarding the logs, all logs lies in this path `~/Library/Logs/rancher-desktop`.

**Comment by drenze-athene** [comment]
> Regarding the logs, all logs lies in this path `~/Library/Logs/rancher-desktop`.

Ya - that's where I found what I posted. There was nothing in there specifically from the installation, and what I posted was from the first startup after I installed, which is when it tried to pull everything down.

**Comment by mook-as** [comment]
Hi! Thanks for the help so far.

The relevant logs would have been in `k8s.log`; specifically, from [here](https://github.com/rancher-sandbox/rancher-desktop/blob/v1.0.0-beta.1/src/k8s-engine/k3sHelper.ts#L224).  It's failing to fetch the list of Kubernetes versions to `~/Library/Caches/rancher-desktop/k3s-versions.json`.  Once it's manage to fetch it once, it's non-blocking on subsequent startups (because it can just use the last known values).

Given that this _only_ VPN has the issue, I'd have thought it would be something to do MITMing proxies; however, I'd have expected that to also exist on-site…

If possible, could you run `openssl s_client -connect update.k3s.io:443 -servername update.k3s.io -showcerts </dev/null` and attach the certificate chain? From where I am, I see:
```
 0 s:/C=US/ST=California/L=San Francisco/O=Cloudflare, Inc./CN=sni.cloudflaressl.com
   i:/C=US/O=Cloudflare, Inc./CN=Cloudflare Inc ECC CA-3
 1 s:/C=US/O=Cloudflare, Inc./CN=Cloudflare Inc ECC CA-3
   i:/C=IE/O=Baltimore/OU=CyberTrust/CN=Baltimore CyberTrust Root
```
(This works because the first certificate has a subject alt name for `DNS:*.k3s.io, DNS:k3s.io, DNS:sni.cloudflaressl.com`; you can see this by passing the `----- BEGIN CERTIFICATE`… `END CERTIFICATE -----` section to `openssl x509 -noout -text`.)  In particular, I'm interested if the _last_ certificate was from your company (or otherwise has an different trust root).

In the mean time, we need to go figure out why the error isn't being logged correctly…

**Comment by drenze-athene** [comment]
> The relevant logs would have been in `k8s.log`; specifically, from [here](https://github.com/rancher-sandbox/rancher-desktop/blob/v1.0.0-beta.1/src/k8s-engine/k3sHelper.ts#L224). It's failing to fetch the list of Kubernetes versions to `~/Library/Caches/rancher-desktop/k3s-versions.json`. Once it's manage to fetch it once, it's non-blocking on subsequent startups (because it can just use the last known values).
> 

That log was empty at the initial startup, at the point where it failed.

> Given that this _only_ VPN has the issue, I'd have thought it would be something to do MITMing proxies; however, I'd have expected that to also exist on-site…
> 

I may have misstated - I have not tried with VPN at this point; I have **only** tried from on-site. My assumption is that you are correct, this is the result of our MITM proxy.

> If possible, could you run `openssl s_client -connect update.k3s.io:443 -servername update.k3s.io -showcerts </dev/null` and attach the certificate chain?

```
Certificate chain
 0 s:/C=US/ST=California/L=San Francisco/O=Cloudflare, Inc./CN=sni.cloudflaressl.com
   i:/DC=tld/DC=corp/DC=/CN=CA-INTERNAL-SUB-A1
 1 s:/DC=tld/DC=corp/DC=mycompany/CN=CA-INTERNAL-SUB-A1
   i:/CN=CA-INTERNAL-ROOT-A
```

Thanks.

**Comment by mook-as** [comment]
Thanks! As expected, you have a MITM proxy (which is fine — that's a scenario we attempt to support).

We attempt to [import your CA certificates](https://github.com/rancher-sandbox/rancher-desktop/blob/v1.0.0-beta.1/src/main/networking/index.ts#L57) on startup, before the failure happened.  The code basically runs `/usr/bin/security find-certificate -a -p` (and again with `/System/Library/Keychains/SystemRootCertificates.keychain`).

Could you check that `CA-INTERNAL-ROOT-A` (or whichever CA cert is in use here) shows up in the list? That should help narrow down if 1) we're not correctly hooking up the CAs, or 2) we didn't find the correct CA to add.

Thanks!

Some notes:
- It might be easier to reproduce this by deleting `k3s-versions.json` to cause it to require a synchronous download on startup (akin to a first run)
- It may be useful to turn on the debugging logs in the `Troubleshooting` page, to get more verbose logging. (Feel free to turn it back off again after you got the logs.)

**Comment by drenze-athene** [comment]
> We attempt to [import your CA certificates](https://github.com/rancher-sandbox/rancher-desktop/blob/v1.0.0-beta.1/src/main/networking/index.ts#L57) on startup, before the failure happened. The code basically runs `/usr/bin/security find-certificate -a -p` (and again with `/System/Library/Keychains/SystemRootCertificates.keychain`).
> 
> Could you check that `CA-INTERNAL-ROOT-A` (or whichever CA cert is in use here) shows up in the list? That should help narrow down if 1) we're not correctly hooking up the CAs, or 2) we didn't find the correct CA to add.
> 

Just did this, what I get is just the raw certs. Is there an easy way to generate a list of which certs are returned from that?

> * It might be easier to reproduce this by deleting `k3s-versions.json` to cause it to require a synchronous download on startup (akin to a first run)
> * It may be useful to turn on the debugging logs in the `Troubleshooting` page, to get more verbose logging. (Feel free to turn it back off again after you got the logs.)

All logs generated by the startup are attached. Github wouldn't let me attach empty files, so here's a screenshot, so you can see which logs are empty:

![image](https://user-images.githubusercontent.com/81439124/150149475-6f26c266-120b-4486-8f7b-1dc910239844.png)

[lima.ha.stderr.log](https://github.com/rancher-sandbox/rancher-desktop/files/7897812/lima.ha.stderr.log)
[lima.ha.stdout.log](https://github.com/rancher-sandbox/rancher-desktop/files/7897813/lima.ha.stdout.log)
[lima.serial.log](https://github.com/rancher-sandbox/rancher-desktop/files/7897814/lima.serial.log)
[update.log](https://github.com/rancher-sandbox/rancher-desktop/files/7897815/update.log)

**Comment by mook-as** [comment]
> > We attempt to [import your CA certificates](https://github.com/rancher-sandbox/rancher-desktop/blob/v1.0.0-beta.1/src/main/networking/index.ts#L57) on startup, before the failure happened. The code basically runs `/usr/bin/security find-certificate -a -p` (and again with `/System/Library/Keychains/SystemRootCertificates.keychain`).
> > Could you check that `CA-INTERNAL-ROOT-A` (or whichever CA cert is in use here) shows up in the list? That should help narrow down if 1) we're not correctly hooking up the CAs, or 2) we didn't find the correct CA to add.
> 
> Just did this, what I get is just the raw certs. Is there an easy way to generate a list of which certs are returned from that?

Maybe just print all the certs? Something like
```sh
/usr/bin/security find-certificate -a -p | awk -v cmd='openssl x509 -noout -text' '/BEGIN/ { close(cmd) } { print | cmd } ' | grep Subject:
```
(Based on [this](https://unix.stackexchange.com/a/97252)).

It would also be good to check `Keychain Access` to see where the desired CA is.
> All logs generated by the startup are attached. Github wouldn't let me attach empty files, so here's a screenshot, so you can see which logs are empty:

Hmm, frustrating that the relevant logs are not available (not sure why; need to investigate separately).

**Comment by mook-as** [comment]
Hmm, the first command had `Subject: CN=CA-INTERNAL-ROOT-A`, so it _should_ have tried to install the CA.  That is, it should be in `/usr/local/share/ca-certificates` in the VM (probably as `rd-1.crt` or something, but the order isn't guaranteed, I think.)

Not sure if you've found the instructions, but to get into the VM, you need to run (from a shell):
```sh
LIMA_HOME="$HOME/Library/Application Support/rancher-desktop/lima" '/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/lima/bin/limactl' shell 0
```
(Adjust path to the binary as necessary.)

At least that rules out the cert being in some unexpected store…

**Comment by drenze-athene** [comment]
> Hmm, the first command had `Subject: CN=CA-INTERNAL-ROOT-A`, so it _should_ have tried to install the CA. That is, it should be in `/usr/local/share/ca-certificates` in the VM (probably as `rd-1.crt` or something, but the order isn't guaranteed, I think.)
> 
> Not sure if you've found the instructions, but to get into the VM, you need to run (from a shell):
> 
> ```shell
> LIMA_HOME="$HOME/Library/Application Support/rancher-desktop/lima" '/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/lima/bin/limactl' shell 0
> ```
> 
> (Adjust path to the binary as necessary.)
> 
> At least that rules out the cert being in some unexpected store…

You are correct. Connecting to the VM, doing `sudo -i` and executing this command:

```bash
for c in *.crt; do printf "$c: " && openssl x509 -noout -subject -in $c; done
```

in `/usr/local/share/ca-certificates` shows the following:

```bash
rd-1.crt: subject=CN = CA-INTERNAL-ROOT-A
```

...so it's there.

**Comment by drenze-athene** [comment]
An additional question: At what point does RD attempt to connect to `https://update.k3s.io/v1-release/channels`? This alert comes up before the RD window even renders, almost as if that's the first thing that's happening. It's as if the UI is checking for this update before the VM even attempts to start.

**Comment by gaktive** [comment]
Hi @drenze-athene. Looking over recent work around this area, we did a fix tied to certificates in https://github.com/rancher-sandbox/rancher-desktop/pull/1613 that was included as part of 1.1.0.

If you're in a position to do so, are you able to try this again with a more recent version of Rancher Desktop (latest: 1.3.0 with another version due soon) and see if you still have the same problem?

**Comment by drenze-athene** [comment]
I have 1.3.0 installed and have had no issues since 1.1.0.

Douglas J. Renze
INFRASTRUCTURE SENIOR ENGINEER

From: Gary A Korhonen ***@***.***>
Date: Tuesday, May 17, 2022 at 4:09 PM
To: rancher-sandbox/rancher-desktop ***@***.***>
Cc: Douglas Renze ***@***.***>, Mention ***@***.***>
Subject: Re: [rancher-sandbox/rancher-desktop] There was an unknown error starting Kubernetes: FetchError: request to https://update.k3s.io/v1-release/channels failed, reason: certificate signature failure (Issue #1261)

ATTENTION: External Email - Be Suspicious of Attachments, Links and Requests for Login Information.

Hi @drenze-athene<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdrenze-athene&data=05%7C01%7Cdrenze%40athene.com%7C349850546b3f43c64e1708da38498861%7C70559d03915449e19e59be933c7f8440%7C0%7C0%7C637884185752145236%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=yQHEEOj%2FEyHM1cai3kVwgFHFkPd%2BB1IDx6%2FHf3S%2B60U%3D&reserved=0>. Looking over recent work around this area, we did a fix tied to certificates in #1613<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Francher-sandbox%2Francher-desktop%2Fpull%2F1613&data=05%7C01%7Cdrenze%40athene.com%7C349850546b3f43c64e1708da38498861%7C70559d03915449e19e59be933c7f8440%7C0%7C0%7C637884185752145236%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=koOyxrd1z2aINHEJDHd0LibO8%2BjePQlFQ6GDFJ%2B4DCw%3D&reserved=0> that was included as part of 1.1.0.

If you're in a position to do so, are you able to try this again with a more recent version of Rancher Desktop (latest: 1.3.0 with another version due soon) and see if you still have the same problem?

—
Reply to this email directly, view it on GitHub<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Francher-sandbox%2Francher-desktop%2Fissues%2F1261%23issuecomment-1129319791&data=05%7C01%7Cdrenze%40athene.com%7C349850546b3f43c64e1708da38498861%7C70559d03915449e19e59be933c7f8440%7C0%7C0%7C637884185752145236%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=v5PCz2bGvulvbc99nt9KIE2bLIDETH3czx%2BSHLgPIyA%3D&reserved=0>, or unsubscribe<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FATNKTFBCUU7O4MK4LXMGRK3VKQDITANCNFSM5MHOXMGA&data=05%7C01%7Cdrenze%40athene.com%7C349850546b3f43c64e1708da38498861%7C70559d03915449e19e59be933c7f8440%7C0%7C0%7C637884185752145236%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=L9VweqqQpCCGmJR%2FBU7gwSbH6z0MlwuJOXXDd8hocdE%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Electronic communication sent through the internet is not secure and its delivery is not guaranteed. This e-mail communication and any attachments may contain confidential and privileged information for the use of the designated recipients named above. If you are not the intended recipient, you are hereby notified that you have received this communication in error and that any review, disclosure, dissemination, distribution or copying of it or its contents is prohibited. If you have received this communication in error, please notify me immediately by replying to this message and deleting it from your computer. All annuity products and services are offered through Athene Annuity & Life Assurance Company and Athene Annuity and Life Company, in all states except New York, and in New York through Athene Annuity & Life Assurance Company of New York and Athene Life Insurance Company of New York. All investment advisory services are rendered solely through Apollo Insurance Solutions Group LLC. None of the information contained herein should be construed as an offer or sale of any security, product, or service of Apollo Insurance Solutions Group LLC. Past performance is not indicative of future success.
